### PR TITLE
Instrain profile

### PR DESCRIPTION
**Pull request by MrOlm**
## PR checklist

- [X] This pull request adds the module "instrain/profile" to nf-core/modules
- [X] If you've fixed a bug or added code that should be tested, add tests!
- [X] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [X] If necessary, include test data in your PR (no new test data required)
- [X] Remove all TODO statements.
- [X] Emit the `versions.yml` file.
- [X] Follow the naming conventions.
- [X] Follow the parameters requirements.
- [X] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [X] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [X] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/instrain/profile/main.nf [hunk 1]
```diff
@@ -0,0 +1,46 @@
+process INSTRAIN_PROFILE {
+    tag "$meta.id"
+    label 'process_high'
+    label 'error_retry'
+    label 'process_long'
```
Comment (jfy133) [inline modules/instrain/profile/main.nf:5]: I'm not sure if adding both the high and long labels will work here, do you know if the order of the multiple labels have an inherant priority order?

File: modules/instrain/profile/main.nf [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    conda (params.enable_conda ? "bioconda::instrain=1.6.1" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/instrain%3A1.6.1--pyhdfd78af_0':
```
Comment (jfy133) [inline modules/instrain/profile/main.nf:9]:
```suggestion
        'https://depot.galaxyproject.org/singularity/instrain:1.6.1--pyhdfd78af_0':
```

Shouldn't this be a `:`? I assume actually it's equivalent but better to have the same as other modules/container IDs in this regard

File: modules/instrain/profile/main.nf (hunk 1 above)
Comment (jfy133) [reply modules/instrain/profile/main.nf:5]: (as in it's the first time I've seen this, so  I genuinely don't know!)

File: modules/instrain/profile/meta.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        e.g. [ id:'test']
+  - bam:
+      type: path
+      description: |
+        Path to .bam file to be profiled
+      pattern: "*.{bam,sam}"
+  - genome_fasta:
+      type: path
+      description: |
+        Path to .fasta file to be profiled; MUST be the .fasta file that was mapped to to create the .bam file
```
Comment (jfy133) [inline modules/instrain/profile/meta.yml:31]:
```suggestion
        Path to .fasta file to be profiled; MUST be the .fasta file that was mapped to to create the .bam file
     pattern: "*.{fasta,fna,fa}"
```

File: modules/instrain/profile/meta.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  - genes_fasta:
+      type: path
+      description: |
+        Path to .fna file of genes to be profiled (OPTIONAL)
```
Comment (jfy133) [inline modules/instrain/profile/meta.yml:35]:
```suggestion
        Path to .fna file of genes to be profiled (OPTIONAL)
     pattern: "*.{fasta,fna,fa}"
```

File: modules/instrain/profile/meta.yml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  - stb_file:
+      type: path
+      description: |
+        Path to .stb (scaffold to bin) file to be profiled (OPTIONAL)
```
Comment (jfy133) [inline modules/instrain/profile/meta.yml:39]:
```suggestion
        Path to .stb (scaffold to bin) file to be profiled (OPTIONAL)
     pattern: "*.stb"
```

File: modules/instrain/profile/meta.yml [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - profile:
+      type: path
+      description: InStrain profile folder
```
Comment (jfy133) [inline modules/instrain/profile/meta.yml:53]:
```suggestion
      description: InStrain profile folder
      pattern: "*.IS"
```

Comment (jfy133) [reply modules/instrain/profile/meta.yml:53]: Your output emission in the `main.nf` implies you output files i.e. `.IS` would suggest it's not a folder as it has a file xtension? Could you clarify?

Comment (jfy133) [reply modules/instrain/profile/meta.yml:53]: I tried installing in a fresh conda environment the tool but unfortunately I get loads of python traceback for whatever command I try so I can't check myself exactly (sorry)

File: modules/instrain/profile/main.nf [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def genes_args = genes_fasta != [] ? "-g ${genes_fasta}": ''
+    def stb_args = stb_file != [] ? "-s ${stb_file}": ''
+    """
+    inStrain \\
+        profile \\
+        $bam \\
+        $genome_fasta \\
+        -o ${meta.id}.IS \\
```
Comment (jfy133) [inline modules/instrain/profile/main.nf:35]:
```suggestion
        -o ${prefix}.IS \\
```

Usage of prefix allows the user to specify an alternative to just meta.id

File: modules/instrain/profile/main.nf [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    tuple val(meta), path("*.IS") , emit: profile
+    path "versions.yml"           , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def genes_args = genes_fasta != [] ? "-g ${genes_fasta}": ''
```
Comment (jfy133) [inline modules/instrain/profile/main.nf:28]:
```suggestion
    def genes_args = genes_fasta ? "-g ${genes_fasta}": ''
```

By default just invoking the name of the variable in a condition results in a boolean output, you don't need to define the condition

File: modules/instrain/profile/main.nf [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    def stb_args = stb_file != [] ? "-s ${stb_file}": ''
```
Comment (jfy133) [inline modules/instrain/profile/main.nf:29]:
```suggestion
    def stb_args = stb_file ? "-s ${stb_file}": ''
```

**Review by jfy133 — CHANGES_REQUESTED** [review]
Mostly a few 'clean up' things! but overall looks good :+1:

File: modules/instrain/profile/main.nf (hunk 1 above)
Comment (MrOlm) [reply modules/instrain/profile/main.nf]: I'm not sure, but I ended up just removing the 'error_retry' and 'process_long' labels; I really don't think they're needed anyways

**Comment by MrOlm** [comment]
Thank you for the detailed review @jfy133 ! I've implemented all suggested changes and I'll keep these conventions in mind the next time I push a module

File: modules/instrain/profile/meta.yml (hunk 6 above)
Comment (jfy133) [reply modules/instrain/profile/meta.yml:50]: OK in this case updte the output emit capture and pattern here to include `/` at the end to make this clearer

**Review by jfy133 — APPROVED** [review]
Two suggestionst o make clear output is directories, but otherwise you're good to go to merge once tests pass!

File: modules/instrain/profile/main.nf [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        'quay.io/biocontainers/instrain:1.6.1--pyhdfd78af_0' }"
+
+    input:
+    tuple val(meta), path(bam)
+    path genome_fasta
+    path genes_fasta
+    path stb_file
+
+    output:
+    tuple val(meta), path("*.IS/") , emit: profile
```
Comment (jfy133) [inline modules/instrain/profile/main.nf:17]:
```suggestion
    tuple val(meta), path("*.IS") , emit: profile
```

Ok apparently NXF didn't like that... but wil keep it for meta

**Comment by jfy133** [comment]
You can merge @MrOlm !

**Pull request merged**